## PBS configure 
#PBS -N mutaion
#PBS -j oe
#PBS -q batch
#PBS -S /bin/sh
#PBS -l nodes=1:ppn=20
#PBS -l mem=16G

source /public/home/nieyg/.bash_profile
cd /md01/nieyg/project/lineage_tracing/bulk_ATAC/masked_genome
reference_genome=/md01/nieyg/ref/hard-mask/genome_modify/assembly_hardmask.fa
cat sample.info |while read id;
do echo $id
arr=($id)
sample=${arr[0]}
echo $sample
samtools index ./2_bam/${sample}.sorted.bam 
# Step2: picard remove the replicates
picard MarkDuplicates CREATE_INDEX=true ASSUME_SORTED=true VALIDATION_STRINGENCY=SILENT REMOVE_DUPLICATES=true \
INPUT=./2_bam/${sample}.sorted.bam  \
OUTPUT=./2_bam/${sample}.sorted.rmdup.bam  \
METRICS_FILE=./2_bam/${sample}.sorted.rmdup.metrics

picard AddOrReplaceReadGroups \
      I=./2_bam/${sample}.sorted.rmdup.bam \
      O=./2_bam/${sample}.AddOrReplaceReadGroups.bam \
      SORT_ORDER=coordinate \
      RGID=4 \
      RGLB=lib1 \
      RGPL=ILLUMINA \
      RGPU=unit1 \
      RGSM=${sample}
samtools index ./2_bam/${sample}.AddOrReplaceReadGroups.bam

java -Xmx8g -jar /md01/nieyg/software/GenomeAnalysisTK_3.5-0.jar -R $reference_genome  -T RealignerTargetCreator  -nt 32 -I ./2_bam/${sample}.AddOrReplaceReadGroups.bam -o ./7_mutation/${sample}.target.intervals

java -Xmx100g -jar /md01/nieyg/software/GenomeAnalysisTK_3.5-0.jar -R $reference_genome -T IndelRealigner -filterNoBases -maxReads 10000000  -I ./2_bam/${sample}.AddOrReplaceReadGroups.bam -targetIntervals ./7_mutation/${sample}.target.intervals
 -o ./7_mutation/${sample}.sorted.rmdup.realign.bam

# Step4: mpileup get file for snv calling

samtools mpileup  -l /md01/nieyg/ref/hard-mask/genome_modify/chrM.len -f $reference_genome -q 10 -Q 10 ./7_mutation/${sample}.sorted.rmdup.realign.bam >  ./7_mutation/${sample}.sorted.rmdup.realign.mpileup

# make count table for each allele 每个线粒体位点的测序情况
/public/home/jinxu/bin/pileup_inf_rj.pl ./7_mutation/${sample}.sorted.rmdup.realign.mpileup >  ./7_mutation/${sample}.count

# Step5: varscan call somatic mutation in chrM
varscan  pileup2snp ./7_mutation/${sample}.sorted.rmdup.realign.mpileup --min-var-freq 0.000001  --min-reads2 4 > ./7_mutation/${sample}.snv
